src/Servers/Kestrel/Core/src/Internal/SniOptionsSelector.cs

```diff
@@ -0,0 +1,116 @@
+                HttpsConnectionMiddleware.ConfigureAlpn(sslServerOptions, sniConfig.Protocols ?? fallbackHttpProtocols);
+
+                var clientCertificateMode = sniConfig.ClientCertificateMode ?? fallbackOptions.ClientCertificateMode;
+
+                if (clientCertificateMode != ClientCertificateMode.NoCertificate)
+                {
+                    sslServerOptions.ClientCertificateRequired = true;
+                    sslServerOptions.RemoteCertificateValidationCallback = (sender, certificate, chain, sslPolicyErrors) =>
+                        HttpsConnectionMiddleware.RemoteCertificateValidationCallback(
+                            clientCertificateMode, fallbackOptions.ClientCertificateValidation, certificate, chain, sslPolicyErrors);
+                }
+
+                if (name.Equals(wildcardHost, StringComparison.Ordinal))
+                {
+                    _wildcardHostOptions = sslServerOptions;
+                }
+                else if (name.StartsWith(wildcardPrefix, StringComparison.Ordinal))
+                {
+                    _wildcardPrefixOptions.Add((name, sslServerOptions));
+                }
+                else
+                {
+                    _fullNameOptions[name] = sslServerOptions;
+                }
+            }
+        }
+
+        public SslServerAuthenticationOptions GetOptions(string serverName)
+        {
+            SslServerAuthenticationOptions options = null;
+
+            if (!string.IsNullOrEmpty(serverName))
+            {
+                if (_fullNameOptions.TryGetValue(serverName, out options))
+                {
+                    return options;
+                }
+
+                var matchedNameLength = 0;
+                ReadOnlySpan<char> serverNameSpan = serverName;
+
+                foreach (var (nameCandidate, optionsCandidate) in _wildcardPrefixOptions)
+                {
+                    ReadOnlySpan<char> nameCandidateSpan = nameCandidate;
+
+                    // Note that we only slice off the `*`. We want to match the leading `.` also.
+                    if (serverNameSpan.EndsWith(nameCandidateSpan.Slice(wildcardHost.Length), StringComparison.OrdinalIgnoreCase)
+                        && nameCandidateSpan.Length > matchedNameLength)
+                    {
+                        matchedNameLength = nameCandidateSpan.Length;
+                        options = optionsCandidate;
+                    }
+                }
+            }
+
+            options ??= _wildcardHostOptions;
+
+            if (options is null)
+            {
+                if (serverName is null)
+                {
+                    throw new AuthenticationException(CoreStrings.FormatSniNotConfiguredToAllowNoServerName(_endpointName));
+                }
+                else
+                {
+                    throw new AuthenticationException(CoreStrings.FormatSniNotConfiguredForServerName(serverName, _endpointName));
```

Comment:
@Tratcher One unfortunate thing about this, is that you can't use an empty wildcard SNI section (e.g. `"*": {}`) to get endpoint-default behavior for non-matches, because IConfigurationSection.GetChildren() doesn't return empty children. You need at least one sub-property 😦

Reply:
Should it be an array instead? Then that one property would be the name.